High resolution scrolling and extended mouse wheel reports

morgannewellsun opened the pull request:
## Description

This PR is for two closely related features (high resolution scrolling, and extended mouse wheel reports), as well as associated documentation.

Four settings macros/flags are added:

* POINTING_DEVICE_HIGH_RESOLUTION_SCROLL_ENABLE
* POINTING_DEVICE_HIGH_RESOLUTION_SCROLL_MULTIPLIER
* POINTING_DEVICE_HIGH_RESOLUTION_SCROLL_EXPONENT
* WHEEL_EXTENDED_REPORT

The first three modify QMK's mouse report descriptor to enable high resolution scrolling. The fourth enables extended wheel reports, which work just like extended mouse reports. Extended wheel reports are more-or-less necessary for high resolution scrolling to work properly, because otherwise the wheel reports are typically too big to fit in an `int8_t`.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/qmk/qmk_firmware/issues/17585

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

IsaacElenbaas commented on docs/features/pointing_device.md:
```diff
@@ -616,6 +643,249 @@ layer_state_t layer_state_set_user(layer_state_t state) {
+
+### Advanced High Resolution Drag Scroll
+
+This implementation of high resolution drag-scroll implements several features: throttling, rescaling, smoothing, and axis-snapping. As in the previous example, make sure to add `#define WHEEL_EXTENDED_REPORT` to `config.h` in order to avoid overflow errors.
```
This is a very large example section and is much more than "show how to interface with QMK." I can definitely see the value of an existing fully-featured implementation for something as complex as comfortable scrolling with a ball, but the docs are not the place for it. Best-case scenario this gets copied however many times and some are out of date, have extra features added, or break.

IMO this should go in something similar to a `process_record` function but for `pointing_device_task`. We currently have `pointing_device_task_kb` -> `pointing_device_task_user`, while `process_record` calls the similar `process_record_kb` -> `process_record_user` - it shouldn't be too messy to mirror [`process_record`](https://github.com/qmk/qmk_firmware/blob/master/quantum/quantum.c#L266) for [pointing devices.](https://github.com/qmk/qmk_firmware/blob/master/quantum/pointing_device/pointing_device.c#L288) Then this feature can be expanded and options added and such as well when people need something new for themselves.

morgannewellsun replied:
I don't think I 100% understand what you're suggesting, but I'll try my best:

I see that there's a function `pointing_device_task` defined in `pointing_device.c`. This function calls several other functions (e.g. `pointing_device_adjust_by_defines` and `pointing_device_task_kb`) to sequentially process the mouse report prior to handing it off to be sent to the host computer. I think it would be sensible to add a function `pointing_device_drag_scroll` to this "pipeline", which would provide a default implementation of drag scroll. 

Ideally, this function would automatically adjust its behavior based on whether high resolution scrolling is enabled. You could also use compiler flags to control aspects of behavior such as throttling, rescaling, smoothing, axis-snapping, etc. There would also need to be a set of functions `drag_scroll_on`, `drag_scroll_off`, `drag_scroll_toggle`, and `is_drag_scroll_on`.

Does this sound consistent with what you're proposing?

IsaacElenbaas replied:
Yes, exactly - just adding it to the baby `pointing_device_task` chain in the same way we do for the massive `process_record_quantum` keyboard one. We don't have the former formalized as being the same design pattern AFAIK, but it looks like that was the intent to me.

Moving that to a separate PR as you are doing is probably best. The 60Hz limitation can be a separate one (it is not necessary on Linux and really a driver/application issue. . . mice can report cursor movement at 10kHz) which makes doc examples not necessary at all. High-res scrolling itself doesn't need anything special.

morgannewellsun commented:
A few changes:

- Removed drag scroll examples from docs; I'll make a separate PR for a "default drag scroll implementation".
- Added ability for keyboard to parse resolution multiplier feature reports, at least on ChibiOS and LUFA. Not sure how to do it on the other platforms.

IsaacElenbaas commented:
Also just wanted to say this is a great first PR :) thank you for putting in all of the effort!

drashna commented:
No VUSB support?

morgannewellsun commented:
> No VUSB support?

Oh my god there's separate USB HID descriptors in VUSB.

Earlier, I tried poking around at the VUSB code that handles USBRQ_HID_SET_REPORT. However, I was a bit confused - in ChibiOS and LUFA there's stuff there to deal with shared endpoints and NKRO, but not so in VUSB. I guess maybe that has to do with the fact that VUSB has its own USB HID descriptors.

On a different note, I'd like to support VUSB, but I don't have any keyboards that use VUSB, so I'm not sure I'd be able to test any code I write... not sure how to get past this roadblock on my own.

morgannewellsun commented:
Also, I'd like to add that I've got the code ready for a default drag-scroll implementation hooked into `pointing_device_task`. However, the code depends on `is_hires_scroll_on` - which is part of this PR - so if I make a separate PR for drag-scroll, it'll be dependent on this PR... I'd appreciate any advice.

morgannewellsun commented:
I think I've managed to bodge together the appropriate modifications to VUSB. 

At this point, I haven't tested using LUFA or VUSB - only ChibiOS. I don't really know how I would go about testing on non-ChibiOS platforms, because all my keyboards use ChibiOS. If someone could test this out for me, that would be really helpful.

drashna commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -155,6 +158,12 @@ __attribute__((weak)) void pointing_device_init(void) {
 #endif
     }
+#ifdef POINTING_DEVICE_HIRES_SCROLL_ENABLE
+    hires_scroll_resolution = POINTING_DEVICE_HIRES_SCROLL_MULTIPLIER;
```
This isn't defined previously in this file.  So, it will error out. Same with the exponent define.

morgannewellsun commented on docs/features/pointing_device.md:
```diff
@@ -394,6 +394,7 @@ Ideally, new sensor hardware should be added to `drivers/sensors/` and `quantum/
 | Setting                                        | Description                                                                                                                      | Default       |
 | ---------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- | ------------- |
 | `MOUSE_EXTENDED_REPORT`                        | (Optional) Enables support for extended mouse reports. (-32767 to 32767, instead of just -127 to 127).                           | _not defined_ |
+| `WHEEL_EXTENDED_REPORT`                        | (Optional) Enables support for extended wheel reports. (-32767 to 32767, instead of just -127 to 127).                           | _not defined_ |
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -377,28 +390,28 @@ void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
 
 /**
- * @brief clamps int16_t to int8_t
+ * @brief clamps int16_t to int8_t, or int32_t to int16_t
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -377,28 +390,28 @@ void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
- * @param[in] int16_t value
- * @return int8_t clamped value
+ * @param[in] hv_clamp_range_t value
+ * @return mouse_hv_report_t clamped value
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -377,28 +390,28 @@ void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
+    if (value < HV_REPORT_MIN) {
+        return HV_REPORT_MIN;
+    } else if (value > HV_REPORT_MAX) {
+        return HV_REPORT_MAX;
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -377,28 +390,28 @@ void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
 
 /**
- * @brief clamps int16_t to int8_t
+ * @brief clamps int16_t to int8_t, or int32_t to int16_t
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -377,28 +390,28 @@ void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
+ * @brief clamps int16_t to int8_t, or int32_t to int16_t
  *
- * @param[in] clamp_range_t value
+ * @param[in] xy_clamp_range_t value
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -377,28 +390,28 @@ void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
  * @return mouse_xy_report_t clamped value
  */
-static inline mouse_xy_report_t pointing_device_xy_clamp(clamp_range_t value) {
+static inline mouse_xy_report_t pointing_device_xy_clamp(xy_clamp_range_t value) {
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.c:
```diff
@@ -419,10 +432,10 @@ static inline mouse_xy_report_t pointing_device_xy_clamp(clamp_range_t value) {
+    left_report.x = pointing_device_xy_clamp((xy_clamp_range_t)left_report.x + right_report.x);
+    left_report.y = pointing_device_xy_clamp((xy_clamp_range_t)left_report.y + right_report.y);
+    left_report.h = pointing_device_hv_clamp((hv_clamp_range_t)left_report.h + right_report.h);
+    left_report.v = pointing_device_hv_clamp((hv_clamp_range_t)left_report.v + right_report.v);
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.h:
```diff
@@ -95,11 +95,21 @@ typedef enum {
 #    define XY_REPORT_MIN INT16_MIN
 #    define XY_REPORT_MAX INT16_MAX
-typedef int32_t clamp_range_t;
+typedef int32_t xy_clamp_range_t;
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device.h:
```diff
@@ -95,11 +95,21 @@ typedef enum {
+#else
+#    define HV_REPORT_MIN INT8_MIN
+#    define HV_REPORT_MAX INT8_MAX
+typedef int16_t hv_clamp_range_t;
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device_drivers.c:
```diff
@@ -391,7 +391,7 @@ const pointing_device_driver_t pointing_device_driver = {
 #elif defined(POINTING_DEVICE_DRIVER_pimoroni_trackball)
 
-mouse_xy_report_t pimoroni_trackball_adapt_values(clamp_range_t* offset) {
+mouse_xy_report_t pimoroni_trackball_adapt_values(xy_clamp_range_t* offset) {
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on quantum/pointing_device/pointing_device_drivers.c:
```diff
@@ -406,10 +406,10 @@ mouse_xy_report_t pimoroni_trackball_adapt_values(clamp_range_t* offset) {
+    static uint16_t         debounce      = 0;
+    static uint8_t          error_count   = 0;
+    pimoroni_data_t         pimoroni_data = {0};
+    static xy_clamp_range_t x_offset = 0, y_offset = 0;
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on tmk_core/protocol/report.h:
```diff
@@ -199,6 +199,12 @@ typedef int16_t mouse_xy_report_t;
+#else
+typedef int8_t mouse_hv_report_t;
+#endif
+
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on tmk_core/protocol/report.h:
```diff
@@ -210,8 +216,8 @@ typedef struct {
-    int8_t            v;
-    int8_t            h;
+    mouse_hv_report_t v;
+    mouse_hv_report_t h;
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on tmk_core/protocol/usb_descriptor.c:
```diff
@@ -155,21 +155,55 @@ const USB_Descriptor_HIDReport_Datatype_t PROGMEM SharedReport[] = {
+                HID_RI_INPUT(8, HID_IOF_DATA | HID_IOF_VARIABLE | HID_IOF_RELATIVE),
+#    ifdef POINTING_DEVICE_HIRES_SCROLL_ENABLE
+            HID_RI_END_COLLECTION(0),
+#    endif
```
Has been PARTIALLY included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on tmk_core/protocol/vusb/vusb.c:
```diff
@@ -524,23 +578,59 @@ const PROGMEM uchar shared_hid_report[] = {
+    0x26, 0xFF, 0x7F, //     Logical Maximum (32767)
+    0x95, 0x01,       //     Report Count (1)
+    0x75, 0x10,       //     Report Size (16)
+#    endif
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).

morgannewellsun commented on tmk_core/protocol/vusb/vusb.c:
```diff
@@ -524,23 +578,59 @@ const PROGMEM uchar shared_hid_report[] = {
+    0xC0, //   End Collection
+#    endif
+    0xC0, //   End Collection
+    0xC0, // End Collection
```
Has been included in [a separate PR](https://github.com/qmk/qmk_firmware/pull/24422).
